### DISCUSSION TITLE
Auto Layout and Figma-like Edge Editing Demo

### DISCUSSION THREAD

**Discussion by idootop** [xyflow/xyflow#4271] [Show and tell]
Hi everyone, I've created a demo showcasing the powerful auto layout and edge editing features of ReactFlow. If you're looking for a solution to streamline your workflow and enhance your graph editing experience, this demo is for you! 

👉 Check it out: [idootop/reactflow-auto-layout](https://github.com/idootop/reactflow-auto-layout)

🚀 Online demo: [https://reactflow-auto-layout.vercel.app](https://reactflow-auto-layout.vercel.app/)

### 1. Node Auto Layout

- Supports various auto layout algorithms like [Dagre](https://github.com/dagrejs/dagre), [ELK](https://github.com/kieler/elkjs), [D3-hierarchy](https://github.com/d3/d3-hierarchy), [D3-dag](https://github.com/erikbrinkman/d3-dag) and more.
- Enables automatic layout of nodes with dynamic sizing.
- Supports automatic layout for multiple subflows.
- Allows dynamic adjustment of layout direction, node spacing, port sorting, and other layout parameters.

https://github.com/idootop/reactflow-auto-layout/assets/35302658/952f5021-1cd0-49bf-8dd8-b12521e2a7ce

### 2. Edge Auto Routing

- Utilizes the [A\* search algorithm](https://en.wikipedia.org/wiki/A*_search_algorithm) combined with [Manhattan Distance](https://simple.wikipedia.org/wiki/Manhattan_distance) to find the optimal path for edges.
- Ensures minimal overlap and intersections between edges and nodes at both ends.

https://github.com/idootop/reactflow-auto-layout/assets/35302658/ea9a3657-b1d2-47c8-9a13-3727dfc31d48

### 3. Edge Polyline Drag Editing

- Edges are drawn as right-angled polylines with rounded corners.
- Edges consist of control points, and the line segments between control points can be moved by dragging the control handles.
- During dragging, nearby control points and line segments are automatically merged, and new control points can be automatically split out.

https://github.com/idootop/reactflow-auto-layout/assets/35302658/01f1c5c5-f224-4d12-9a31-bca45a0d5a56

**Comment by moklick** [comment]
Hi @idootop! Wow, this is pretty impressive. Thanks for sharing. I will take a closer look asap :)

**Comment by a3ng7n** [comment]
This looks awesome; would totally love to use this. Though I'm finding it difficult to follow - is there a primer by chance on how the code is organized? I'm selfishly interested in the edge routing and drag editing.

**Reply by idootop** [reply]
This demo is divided into several modules based on functionality, most of which can be directly copied and used. Let's break it down:

### Basic Types:

- [src/data/types.ts](https://github.com/idootop/reactflow-auto-layout/blob/4668e58465742be0b082623ccf03060921c7345a/src/data/types.ts): Contains type definitions for node and edge data. Reviewing this will help you understand the rest of the code.

### Node Auto Layout:

- [src/layout/node/algorithms](https://github.com/idootop/reactflow-auto-layout/blob/4668e58465742be0b082623ccf03060921c7345a/src/layout/node/algorithms): Contains implementations of various node layout algorithms.
- [src/layout/useAutoLayout.ts](https://github.com/idootop/reactflow-auto-layout/blob/4668e58465742be0b082623ccf03060921c7345a/src/layout/useAutoLayout.ts): Handles the auto layout process, including logic for dynamically adapting to node sizes.

### Edge Editing Functionality:

- [src/layout/edge/index.ts](https://github.com/idootop/reactflow-auto-layout/blob/4668e58465742be0b082623ccf03060921c7345a/src/layout/edge/index.ts): Start here to explore the control point generation algorithms and logic for drawing rounded corner edge paths.
- [src/layout/edge/algorithms/index.ts](https://github.com/idootop/reactflow-auto-layout/blob/4668e58465742be0b082623ccf03060921c7345a/src/layout/edge/algorithms/index.ts): Core of the edge auto-routing algorithm. Refer to the [LogicFlow 边的绘制与交互](https://juejin.cn/post/6942727734518874142) article for more details.
- [src/components/Edges/EdgeController/index.tsx](https://github.com/idootop/reactflow-auto-layout/blob/4668e58465742be0b082623ccf03060921c7345a/src/components/Edges/EdgeController/index.tsx): Follow this to understand how edge segment drag events are handled.
- [src/components/Edges/EdgeController/smart-edge.ts](https://github.com/idootop/reactflow-auto-layout/blob/4668e58465742be0b082623ccf03060921c7345a/src/components/Edges/EdgeController/smart-edge.ts): Manages logic for edge auto-merging/splitting, similar to Figma.

These are the key modules of the project. While it might seem complex at first, the overall logic is straightforward. If you have any questions, feel free to raise an issue.

Additionally, please note this project is a simple demo to showcase ReactFlow's capabilities for node auto-layout and Figma-like edge editing. It hasn't undergone extensive performance optimization or code cleanup, serving mainly as a reference.

> PS: The code comments are mostly in Chinese, which might be a bit challenging for you :)

**Reply by a3ng7n** [reply]
Thanks so much for the explanation!

**Reply by a3ng7n** [reply]
> PS: The code comments are mostly in Chinese, which might be a bit challenging for you :)

Well maybe I'm fluent in Chinese! (I wish...). I did google translate some of the comments to english to make it easier to read (they seem to be _close_ ) - if it would help I'm happy to publish a PR with them alongside the originals.

Again, thank you for publishing this!

**Reply by idootop** [reply]
Awesome! Your effort to translate the comments is much appreciated—though ChatGPT might be a better option than Google Translate. 😄 I'd be delighted to see your PR with the translated comments. Thanks again for your enthusiasm and support!

**Reply by a3ng7n** [reply]
Pretty sure I got everything, and didn't munge any linting codes, but I only did a cursory glance to make sure they "made sense": https://github.com/idootop/reactflow-auto-layout/pull/1

**Reply by idootop** [reply]
LGTM, merged :) Thank you for submitting the PR and taking the time to translate the code comments into English.

**Reply by idootop** [reply]
hey @a3ng7n, Just wanted to let you know that I've polished up the English translation based on your work. Feel free to take a look at the updated translations whenever you have a chance. Thanks again for your contributions!

**Reply by moklick** [reply]
It's getting even better 🤩
